Messaging

cemgunay opened the pull request:
MESSAGINGGGGGGGG

this fixes #25 

Ok so this ended up being way bigger of a branch than I thought but it's still pretty straightforward I'll try explain best as I can and how you can play around with it.

### Inbox, Conversation, Messages
- There are 2 different "inboxes", one is the tenant inbox and the other is the subtenant inbox.
- The tenant inbox is accessed from `/host/inbox` and the subtenant inbox is accessed from `/inbox.`
- This is because a user can be both a tenant and a subtenant.
- So if there is **user1** who is the owner of **listing1** and there is a **user2** who wants to message **user1** about **listing1**
  - On the listing page (`/listing/[listingId]`) under "Subletted by **user1**" it will say "contact host" and **user2** can click that to send a message. 
    - This checks if there is already a direct conversation open between **user1** and **user2** about this listing and if there is it will update the existing conversation with the new message instead of creating a new conversation)
- A conversation is created with the 2 users and the listingId. The users however have "roles" in this conversation where in this case **user1** has role tenant and **user2** has role subtenant.
- Now if **user1** goes to `/host/inbox` they will see this conversation but if they go to `/inbox` they will not see this conversation, and vice versa for `user2`.
- In the inbox routes, users can also create group chats with name and multiple users.
- For group chats everyone gets the role "subtenant", unless it is created from the `/host/inbox` route where in that case the person creating it gets the "tenant" role
- Users can be added to existing conversations (either direct conversations or group chats) and a system message is sent to say that **userX** has added **userY**
- Messages can be either text or image (uses cloudinary client)
- Conversations can also be deleted and this simply just does a "soft delete" for the user and if the other users send a message the chat comes back to the user that had it "soft deleted" but with the past messages gone and only the ones after their "soft delete date". They can do this over and over.

### Real Time
- Abused the shit out of Pusher with this messaging.
- Users can see new and updated conversations in the inbox (`/inbox `and `/host/inbox`) in real time
- Users can see new messages and updated messages (Read Receipts) in the conversation in real time.
- Users can see new people being added to group chats in real time

### Best way to play around...
1. Have 2 browsers open with different users signed in (cem2 and cem1 already have messages or cem2 and nathanclairmonte@outlook.com)
2. Create a direct message from "contact host" button in listing or just create a group chat from inbox
3. See the conversation get created on other browser
4. Send messages/images back and forth
5. See the conversation if you are in `inbox` get updated real time or if you are in messages (`/inbox/conversation/[conversationId]`) see the messages get updated in real time with read receipts
6. Add users to existing conversations to see the system messages
7. See the drawer update with new emails when new users are added in real time

note: cant remove users after adding them LOL but im too lazy to add that for this version of app I think (wouldnt be that hard with adding conversationOwner when creating but whatever can always add later). Also cant leave a group chat LOL.

LMK IF ANYTHING BREAKS

also change NEXT_PUBLIC_CLOUD_NAME to NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME in .env

nathanclairmonte commented:
Looks amazingggg bro this is really good stuff!! With the real-time it's wild how instantaneous it happens. I honestly don't have any like concrete changes I think we should make tbh! Here are few small things I noticed:
- On the /inbox and /host/inbox pages the real-time updates for a new message don't happen until that browser tab is in focus again.
  - For example, let's say I have two browser tabs currently open side by side. On the left I'm logged in as nathanclairmonte@outlook.com and I'm on the /host/inbox page. On the right I'm logged in as cem2 and I'm on the conversation page for the convo with nathanclairmonte@outlook.com.
  - If I send a message from the right browser tab (at which point the right browser tab will now be the one in focus), that message doesn't show up in the host inbox page on the left browser tab until I click somewhere on that page (i.e. bring the left browser tab back into focus again). I hope this makes sense!
  - Honestly I don't see this as a huge issue because a user will only ever be using like one version of the app at a time, and therefore if they're on any of the inbox pages it _should_ always be focused.
  - But it is a disparity, because the messages themselves inside the convo page _do_ update without the browser being in focus. (i.e. imagine the exact same example as above, but instead of being on the inbox page on the left browser tab I'm instead in the conversation page. then the real-time updates work without the left browser tab being in focus). Again not a huge issue but I can't quite figure out why it works without browser focus for the conversation page and not for the inbox pages.
- I'm getting some random refreshes sometimes when I click from one tab to the other. I like scoured all the code to see if I could pinpoint something that is causing it but I can't 😩. Maybe something to do with Pusher or the SWR stuff? I'm not too sure, will do some subsequent research

And that's it really! Here's what I tested out that is working perfectly:
- Message sending ofc. No issues at all, all messages went through and UI updates for a message send are perfect. Real-time and everything.
- Image sending works great. Speed of the send is insane too even for relatively large images (I used an image I was gonna set as my laptop background, kinda thinking it might be too large lol, but it sent perfectly fine!)
- "Seen" functionality is working perfectly also, for both text and image messages
- "Contact host" functionality working perfectly
- Creating a new group works perfectly
- Deleting convos also works perfectly (I like that deleting a group convo doesn't take you out the group. new messages in that group will still come thru)

cemgunay commented:
Fixed, had to do with incorrect update message handling for the inbox updates and fixed the random refreshes by removing loading screen on SWR validation. I don't remember why I put that in there I hope it was by accident LOL
